File: 36-valid-sudoku/36-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int n = 9;
        int[][] rows = new int[n][n];
        int[][] cols = new int[n][n];
        int[][] boxes = new int[n][n];
        
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                if (board[r][c] == '.') {
                    continue;
                }
                int num = board[r][c] - '1';
                
                if (rows[r][num]==1) {
                    return false;
                }
                rows[r][num]++;
                
                if (cols[c][num]==1) {
                    return false;
                }
                cols[c][num]++;
                
                int box = (r/3)*3 + c/3;
                if (boxes[box][num]==1) {
                    return false;
                }
                boxes[box][num]++;
            }
        }
        return true;
    }
}

File: 36-valid-sudoku/36-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        const int n = 9;
        bool rows[n][n] = {false};
        bool cols[n][n] = {false};
        bool boxes[n][n] = {false};
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                int digit = board[i][j]-'0'-1;
                int box = i/3 + (j/3)*3;
                if (rows[i][digit] || cols[j][digit] || boxes[box][digit]) {
                    return false;
                }
                rows[i][digit] = true;
                cols[j][digit] = true;
                boxes[box][digit] = true;
            }
        }
        return true;
    }
};